Add support for tracing vmap in pre-dispatch export

tugsbayasgalan opened the pull request:
Summary: ONNX team and recent transformer upgrade ran into this error and we also ran into during our export benchmarking. This diff makes it possible to trace through vmap implementation in pre-dispatch IR. Note that we don't support serializing functorch ops in pre-dispatch IR and in the future, we should desugar them to post-grad ops.

The implementation strategy is:
1. We add python wrappers around vmap APIs so that we attach custom torch function handler that is only on during non-strict export. The reason is we don't want to add this to default torch_function handler because it will break BC. 
2. Some dynamo changes to make sure it picks up new python wrapper APIs. The reason is when we do strict export, we need to re-materialize these APIs in pre-dispatch IR from torch IR. We can avoid this by special casing in dynamo for export to proxy different API calls but i feel that is too much chaos because you need to be able to proxy 2 different variants of same vmap API. 

Test Plan: CI

Differential Revision: D75623875




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

bdhirsh commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1371,6 +1372,38 @@ def __torch_function__(
             return node
             # Don't actually run the function! We just want to trace the calls
             # into a graph. We don't actualy want to change global autograd state.
+        if func in [torch._C._functorch._add_batch_dim, torch._C._functorch._remove_batch_dim]:
```
this is a very high level take but i'm surprised that so many changes were needed. The naive way I would have implemented this is:

(1) find the location that each of these ~5 APIs is implemented and add a torch_function handler to each (a few might already have torch_function handlers)

(2) Take the existing `TorchFunctionMode` we use in pre-dispatch tracing and tweak it to add these ops to its allow-list of what to proxy

bdhirsh commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1371,6 +1372,38 @@ def __torch_function__(
+            # Don't actually run the function! We just want to trace the calls
+            # into a graph. We don't actualy want to change global functorch state.
+
+            def fetch_proxy(x):
```
hmm this feels like a lot of boilerplate:

(1) are these functions below identical to the ones we're calling from inside the `ProxyTorchDispatchMode`? it might be cleaner to split this into a shared util function or two that we can call from both places

(2) can we just unconditionally run the logic to grab input proxies / set output proxies using `pytree.tree_map_only(torch.Tensor)`? that should automatically handle ops that do/do not have tensor inputs or outputs

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -2027,6 +2027,10 @@ def _export_for_training(
 
+    # Load the functorch decomp here 
+    from torch._functorch.vmap import lazy_load_decompositions
+    lazy_load_decompositions()
```
i remember you needed to add some code to lazily load CIA decomps into the pytorch dispatcher. should this lazy loading happen in a similar location?

zou3519 replied:
I'm confused, the first time vmap gets called the decomps will be lazily loaded. You don't need that here if that's the case, right?

tugsbayasgalan replied:
If it is non-strict export, this will be the first time we load it.

bdhirsh commented on torch/_functorch/apis.py:
```diff
@@ -191,7 +191,7 @@ def vmap(
     """
     from torch._dynamo import is_compiling
-
+    
```
unnecessary whitespace change?

bdhirsh commented on torch/_dynamo/variables/builder.py:
```diff
@@ -2852,6 +2852,9 @@ def handle_traced_output(example_value, tx, proxy, options, subclass_type, targe
     elif isinstance(example_value, float) or proxy.node.target in ["hex", "__round__"]:
         set_example_value(proxy.node, example_value)
         return ConstantVariable.create(example_value, **options)
+    elif isinstance(example_value, int) and proxy.node.target in [torch._C._functorch._vmap_increment_nesting, torch._C._functorch._vmap_decrement_nesting]:
```
hmm also - I see you added the first two API's to `trace_rules` - do we need both changes to properly support these APIs in dynamo?

zou3519 reviewed (commented):
Our conclusion from the meeting on Tuesday was:
1) Yes, we're going to put all the API calls into the graph
2) We should only interpose on these API calls when non-strict export is on. These shouldn't go through regular torch_function, because they are private APIs.
3) It might be easier to do this by creating a python function wrapper around e.g. add_batch_dim, and then putting the torch_function handler and export checks into said function. This may require you to add these new functions to Dynamo skiplists to not break the Dynamo side of things

ezyang commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -5539,6 +5540,7 @@ def forward(self, L_x_: "f32[4, 3]", L_y_: "f32[3, 4]"):
         return (unflatten, unflatten_1)
 """,
         )
+        print("ACT", actual)
```
print

ezyang commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -5461,6 +5461,7 @@ def wrapper_fn(x, y):
             return
 
         actual = normalize_gm(wrapped_gm.print_readable(print_output=False))
+        print(actual)
```
print

ezyang commented on torch/csrc/functorch/init.cpp:
```diff
@@ -7,6 +7,8 @@
 #include <ATen/WrapDimUtils.h>
 #include <torch/csrc/functorch/init.h>
+#include <torch/csrc/utils/disable_torch_function.h>
+#include <torch/csrc/utils/python_arg_parser.h>
```
??

tugsbayasgalan replied:
Not needed.

ezyang commented on torch/_export/verifier.py:
```diff
@@ -222,6 +222,10 @@ def _allowed_op_types() -> tuple[type[Any], ...]:
+                torch._functorch.vmap._add_batch_dim_python_wrapper,
+                torch._functorch.vmap._remove_batch_dim_python_wrapper,
+                torch._functorch.vmap._vmap_increment_nesting_python_wrapper,
+                torch._functorch.vmap._vmap_decrement_nesting_python_wrapper,
```
So what does it take to remove this allowlist?

tugsbayasgalan replied:
I think similar story as subclasses serialization. We will eventually just joint IR trace the parts that are not seriaoizable and put Autograd HOP.

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -817,6 +817,44 @@ def _maybe_record_pointwise_barrier(
 
+def _fetch_proxies_and_all_constant_flag(
+    flat_args_kwargs: list[object], tracer: Tracer
+) -> tuple[list[object], tuple[object, ...], bool]:
```
doc me?

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -869,27 +907,8 @@ def can_handle_tensor(x: Tensor) -> bool:
-        # this can happen
-        and not any(isinstance(x, py_sym_types) for x in flat_args_kwargs)
+    f_flat_args_kwargs, proxy_flat_args_kwargs, all_constant = (
+        _fetch_proxies_and_all_constant_flag(flat_args_kwargs, tracer)
```
Is this code motion only?

ezyang commented on torch/_export/utils.py:
```diff
@@ -211,6 +212,29 @@ def _getattr(model: torch.fx.GraphModule, attr_name: str):
+    if not torch._C._is_torch_function_mode_enabled():
+        return None
+
+    torch_function_mode_stack = torch.overrides._get_current_function_mode_stack()
```
There's a lot of these stack search utilities all around everywhere, maybe not this PR but any chance we can centralize them all in overrides? So it's easier to see all of them in one place.

ezyang commented on torch/_export/wrappers.py:
```diff
@@ -186,23 +187,12 @@ def wrapper(*args, **kwargs):
             return
 
-        mode = pre_dispatch_tf_modes[0]
+        assert isinstance(mode, PreDispatchTorchFunctionMode)
```
Code motion only?

tugsbayasgalan replied:
Not needed

ezyang commented on torch/_functorch/vmap.py:
```diff
@@ -149,14 +149,78 @@ def _process_batched_inputs(
 
 # Creates BatchedTensors for every Tensor in arg that should be batched.
 # Returns the (potentially) batched arguments and the batch_size.
+def _add_batch_dim_python_wrapper(self, batch_dim, level):
```
Would like docs explaining why these exist here.

ezyang commented on torch/_functorch/vmap.py:
```diff
@@ -149,14 +149,78 @@ def _process_batched_inputs(
+
+    tracer = mode.tracer
+
+    _, proxies, _ = _fetch_proxies_and_all_constant_flag(
```
So.... please explain the BC problem here?

ezyang replied:
Can you elaborate more on what broke when you just used `handle_torch_function`? In general we expect our torch function dispatch modes to be robust to new operators showing up

zou3519 replied:
+1, let's use handle_torch_function mode than run our own logic here

tugsbayasgalan replied:
Hmm here we are making sure to dispatch to pre-dispatch export torch function mode only.

ezyang commented:
This all seems basically reasonable!

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -5262,9 +5262,9 @@ def forward(self, L_x_: "f32[4, 3]"):
+        results: "f32[12, 4, 3]" = torch._functorch.vmap._remove_batch_dim_python_wrapper(tangents_out_unflatten, 1, 12, 0);  tangents_out_unflatten = None
 
-        _vmap_decrement_nesting = torch._C._functorch._vmap_decrement_nesting();  _vmap_decrement_nesting = None
+        _vmap_decrement_nesting_python_wrapper = torch._functorch.vmap._vmap_decrement_nesting_python_wrapper();  _vmap_decrement_nesting_python_wrapper = None
```
I don't want to go down everything and name them "*_python_wrapper"

tugsbayasgalan replied:
I think the original argument was that we don't want to add tf handler to the C++ function. Are you saying we still define a python wrapper and inside there we just proxy torch._C._functorch._vmap_decrement_nesting?

zou3519 commented on test/export/test_export.py:
```diff
@@ -2401,6 +2401,23 @@ def forward(self, x):
+        ep = export(Vmap(), inputs, {}, dynamic_shapes=dynamic)
+        self.assertTrue(torch.allclose(ep.module()(*inputs), Vmap()(*inputs)))
+
+    @testing.expectedFailureLegacyExportNonStrict  # Old export doesn't work with subclasses
+    @testing.expectedFailureLegacyExportStrict  # Old export doesn't work with subclasses
```
How did you break this in this PR?

tugsbayasgalan replied:
This is like pre dispatch funcitonalization + vmap. Since this path is fully deprecated, we just ignore the failure.

zou3519 commented on test/export/test_export.py:
```diff
@@ -2401,6 +2401,23 @@ def forward(self, x):
+    def test_vmap(self):
+        class Vmap(torch.nn.Module):
+            def forward(self, x, y):
+                f = lambda x, y: x * y + 1  # noqa: E731
+                return torch.vmap(f)(x, y)
```
you probably want something a bit more difficult to vmap than this

zou3519 reviewed (commented):
1) I don't know what is going on in symbolic_shapes
2) I would prefer naming these APIs something less wordy or more meaningful than appending "python_wrapper" to everything
3) we should check that we didn't break AOTAutogradCache, if that was supported in the first place

tugsbayasgalan commented on torch/_dynamo/variables/builder.py:
```diff
@@ -3020,6 +3019,12 @@ def handle_traced_output(example_value, tx, proxy, options, subclass_type, targe
     elif isinstance(example_value, float) or proxy.node.target in ["hex", "__round__"]:
         set_example_value(proxy.node, example_value)
         return ConstantVariable.create(example_value, **options)
+    elif isinstance(example_value, int) and proxy.node.target in [
+        torch._functorch.vmap._vmap_increment_nesting_python_wrapper,
+        torch._functorch.vmap._vmap_decrement_nesting_python_wrapper,
+    ]:
+        set_example_value(proxy.node, example_value)
+        return ConstantVariable.create(example_value, **options)
```
It was just complaining the fx graph was returning non-tensor. I was just cargo culting here.

guilhermeleobas replied:
> Do we guard on the levels?

Yes. The guard is called `FUNCTORCH_STACK_MATCH` and it is called from `VmapIncrementNestingCtxManagerVariable.enter`

@tugsbayasgalan, I didn't get why do you need to wrap the vmap C functions in Python wrappers. Also, there is an if-statement in builder.py that you can add those functions.

https://github.com/pytorch/pytorch/blob/e5a81aa7baaee4d971bdbf3ba3d3af73d052c3f1/torch/_dynamo/variables/builder.py#L2970-L2993

tugsbayasgalan commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -3088,29 +3088,29 @@ def forward(self, L_a_ : torch.SymInt, L_b_ : torch.SymInt, L_c_ : torch.SymInt,
     lazy_load_decompositions = torch._functorch.vmap.lazy_load_decompositions();  lazy_load_decompositions = None
-    _vmap_increment_nesting = torch._C._functorch._vmap_increment_nesting(l_d_, 'error');  _vmap_increment_nesting = None
-    child = torch._C._functorch._add_batch_dim(d, 0, 1);  d = None
+    _vmap_increment_nesting_python_wrapper = torch._functorch.vmap._vmap_increment_nesting_python_wrapper(l_d_, 'error');  _vmap_increment_nesting_python_wrapper = None
```
Ahh good point! Is there a way to check it?

tugsbayasgalan replied:
cc: @jamesjwu

jamesjwu replied:
The concern is that now dynamo is returning a graph with the vmap node under a python wrapper instead of under torch._C, yes?

The best way to test here is to use the test case that this test uses (or one like it) and try saving/loading it with AOTAutogradCache. Something like this: https://github.com/pytorch/pytorch/blob/89654db1abccf7e5f261989a150db4d1619ea2aa/test/dynamo/test_aot_autograd_cache.py#L1144
except with a function that calls torch.vmap.

You can test to make sure whether AOTAutogradCache saves/loads it before/after this change.

tugsbayasgalan replied:
Added!

zou3519 commented on test/dynamo/test_aot_autograd_cache.py:
```diff
@@ -292,6 +292,58 @@ def fn(x, y):
+            vmapped = torch.vmap(f)(x, y)
+            return vmapped.sum(dim=0)
+
+        torch._functorch.config.strict_autograd_cache = True
```
this needs to be in a context manager or something otherwise you will pollute global test state

tugsbayasgalan replied:
Yeah sorry i was debugging sth and forgot to delete

zou3519 commented on torch/_dynamo/trace_rules.py:
```diff
@@ -244,6 +244,8 @@
     "torch._C.set_autocast_xla_dtype": SkipFunctionVariable,
     "torch._C.set_autocast_xla_enabled": SkipFunctionVariable,
     "torch.resize_as_": SkipFunctionVariable,
+    "torch._functorch.vmap._add_batch_dim_in_graph": TorchInGraphFunctionVariable,
+    "torch._functorch.vmap._remove_batch_dim_in_graph": TorchInGraphFunctionVariable,
```
On second thought, I really don't like the special names. Especially if these will be BC-breaking to change.

Can we name these something sane?
- torch._functorch.vmap._add_batch_dim
- torch._functorch.vmap._remove_batch_dim
- torch._functorch.vmap._vmap_increment_nesting
- torch._functorch.vmap._vmap_decrement_nesting

zou3519 commented on torch/_functorch/vmap.py:
```diff
@@ -149,14 +149,48 @@ def _process_batched_inputs(
 # Creates BatchedTensors for every Tensor in arg that should be batched.
 # Returns the (potentially) batched arguments and the batch_size.
+def _add_batch_dim_in_graph(self, batch_dim, level):
+    """
+    Thin wrapper around torch._C._add_batch_dim that is used to proxy in
+    PT2 export/compile fx graph
+    """
```
description should mention non-strict tracing. this all isn't needed for dynamo tracing

zou3519 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1435,6 +1452,24 @@ def __torch_function__(
                 func(*args, **kwargs)
             return node
+
+        if func in [
+            torch._functorch.vmap._add_batch_dim_in_graph,
+            torch._functorch.vmap._remove_batch_dim_in_graph,
+            torch._functorch.vmap._vmap_increment_nesting_in_graph,
+            torch._functorch.vmap._vmap_decrement_nesting_in_graph,
+            torch._functorch.vmap.lazy_load_decompositions,
+        ]:
+            _, proxies, _ = _fetch_proxies_and_all_constant_flag(args, self.tracer)
+            out_proxy = self.tracer.create_proxy(
+                "call_function",
+                func,
+                proxies,
+                {},
+            )
+            res = func(*args, **kwargs)
+            track_tensor_tree(res, out_proxy, constant=None, tracer=self.tracer)
+            return res
```
why do these need to be special cased?

tugsbayasgalan replied:
Because unlike other func's, these functions require non-trivial handling of input and output proxies. By defualt, torch_function handler receives plain objects, not the proxies.

zou3519 reviewed (commented):
1) caching test seems suspicious
2) I would prefer torch._functorch.predispatch.* naming.
3) I don't understand why we need to special case these functions in proxy_tensor.py

zou3519 commented on test/dynamo/test_aot_autograd_cache.py:
```diff
@@ -292,6 +292,58 @@ def fn(x, y):
+        self.assertEqual(x.grad, x2.grad)
+        self.assertEqual(y.grad, y2.grad)
+
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_miss"], 1)
```
nvm I can't read

zou3519 reviewed (approved):
give me some docs about what is going on in proxy_tensor.py

tugsbayasgalan commented on torch/_export/verifier.py:
```diff
@@ -223,6 +223,11 @@ def _allowed_op_types() -> tuple[type[Any], ...]:
+                torch._functorch.predispatch._remove_batch_dim,
+                torch._functorch.predispatch._vmap_increment_nesting,
+                torch._functorch.predispatch._vmap_decrement_nesting,
+                torch._functorch.vmap.lazy_load_decompositions,
```
Yeah need to figure out why. I think devmate forgot to update here but i don't know why it is not catching it in the CI

tugsbayasgalan committed:
Add support for tracing vmap in pre-dispatch export (#154650)

Summary:

ONNX team ran into this error and we also ran into during our export benchmarking. This diff makes it possible to trace through vmap implementation in pre-dispatch IR. Note that we don't support serializing functorch ops in pre-dispatch IR and in the future, we should desugar them to post-grad ops.

Test Plan: CI

Reviewed By: zou3519

Differential Revision: D75623875
